// @matrixai/async-init/dist/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventAsyncInitDestroyed = exports.EventAsyncInitDestroy = exports.EventAsyncInitStopped = exports.EventAsyncInitStop = exports.EventAsyncInitStarted = exports.EventAsyncInitStart = exports.EventAsyncInit = void 0;
const events_1 = require("@matrixai/events");
class EventAsyncInit extends events_1.AbstractEvent {
}
exports.EventAsyncInit = EventAsyncInit;
class EventAsyncInitStart extends EventAsyncInit {
}
exports.EventAsyncInitStart = EventAsyncInitStart;
class EventAsyncInitStarted extends EventAsyncInit {
}
exports.EventAsyncInitStarted = EventAsyncInitStarted;
class EventAsyncInitStop extends EventAsyncInit {
}
exports.EventAsyncInitStop = EventAsyncInitStop;
class EventAsyncInitStopped extends EventAsyncInit {
}
exports.EventAsyncInitStopped = EventAsyncInitStopped;
class EventAsyncInitDestroy extends EventAsyncInit {
}
exports.EventAsyncInitDestroy = EventAsyncInitDestroy;
class EventAsyncInitDestroyed extends EventAsyncInit {
}
exports.EventAsyncInitDestroyed = EventAsyncInitDestroyed;
