// @matrixai/async-init/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initLock = exports.statusP = exports.status = exports.destroyed = exports.running = exports.errors = exports.events = exports.startStop = exports.createDestroy = exports.createDestroyStartStop = exports.types = void 0;
exports.types = __importStar(require("./types"));
exports.createDestroyStartStop = __importStar(require("./CreateDestroyStartStop"));
exports.createDestroy = __importStar(require("./CreateDestroy"));
exports.startStop = __importStar(require("./StartStop"));
exports.events = __importStar(require("./events"));
exports.errors = __importStar(require("./errors"));
var utils_1 = require("./utils");
Object.defineProperty(exports, "running", { enumerable: true, get: function () { return utils_1.running; } });
Object.defineProperty(exports, "destroyed", { enumerable: true, get: function () { return utils_1.destroyed; } });
Object.defineProperty(exports, "status", { enumerable: true, get: function () { return utils_1.status; } });
Object.defineProperty(exports, "statusP", { enumerable: true, get: function () { return utils_1.statusP; } });
Object.defineProperty(exports, "initLock", { enumerable: true, get: function () { return utils_1.initLock; } });
