// @matrixai/quic/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.native = exports.errors = exports.events = exports.utils = exports.QUICStream = exports.QUICConnectionId = exports.QUICConnection = exports.QUICConnectionMap = exports.QUICClient = exports.QUICServer = exports.QUICSocket = void 0;
var QUICSocket_1 = require("./QUICSocket");
Object.defineProperty(exports, "QUICSocket", { enumerable: true, get: function () { return __importDefault(QUICSocket_1).default; } });
var QUICServer_1 = require("./QUICServer");
Object.defineProperty(exports, "QUICServer", { enumerable: true, get: function () { return __importDefault(QUICServer_1).default; } });
var QUICClient_1 = require("./QUICClient");
Object.defineProperty(exports, "QUICClient", { enumerable: true, get: function () { return __importDefault(QUICClient_1).default; } });
var QUICConnectionMap_1 = require("./QUICConnectionMap");
Object.defineProperty(exports, "QUICConnectionMap", { enumerable: true, get: function () { return __importDefault(QUICConnectionMap_1).default; } });
var QUICConnection_1 = require("./QUICConnection");
Object.defineProperty(exports, "QUICConnection", { enumerable: true, get: function () { return __importDefault(QUICConnection_1).default; } });
var QUICConnectionId_1 = require("./QUICConnectionId");
Object.defineProperty(exports, "QUICConnectionId", { enumerable: true, get: function () { return __importDefault(QUICConnectionId_1).default; } });
var QUICStream_1 = require("./QUICStream");
Object.defineProperty(exports, "QUICStream", { enumerable: true, get: function () { return __importDefault(QUICStream_1).default; } });
exports.utils = __importStar(require("./utils"));
exports.events = __importStar(require("./events"));
exports.errors = __importStar(require("./errors"));
exports.native = __importStar(require("./native"));
__exportStar(require("./types"), exports);
