// @matrixai/quic/dist/QUICConnectionId.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class QUICConnectionId extends Uint8Array {
    string;
    /**
     * Decodes from hex string.
     */
    static fromString(idString) {
        const buf = Buffer.from(idString, 'hex');
        return new this(buf.buffer, buf.byteOffset, buf.byteLength);
    }
    /**
     * Decodes as Buffer zero-copy.
     */
    static fromBuffer(idBuffer) {
        return new this(idBuffer.buffer, idBuffer.byteOffset, idBuffer.byteLength);
    }
    constructor(...args) {
        // @ts-ignore: spreading into Uint8Array constructor
        super(...args);
        this.string = this.toBuffer().toString('hex');
    }
    /**
     * Encodes to hex string.
     */
    toString() {
        return this.string;
    }
    /**
     * Encodes as Buffer zero-copy.
     */
    toBuffer() {
        return Buffer.from(this.buffer, this.byteOffset, this.byteLength);
    }
    [Symbol.toPrimitive](_hint) {
        return this.toString();
    }
}
exports.default = QUICConnectionId;
