// @matrixai/events/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEqualOptions = exports.canonicalizeOptions = exports.handleEventError = exports._handleEventError = exports.eventHandled = exports._eventHandled = exports.eventHandlers = exports._eventHandlers = exports.eventTarget = exports._eventTarget = void 0;
/**
 * Symbols prevents name clashes with decorated classes
 */
const _eventTarget = Symbol('_eventTarget');
exports._eventTarget = _eventTarget;
const eventTarget = Symbol('eventTarget');
exports.eventTarget = eventTarget;
const _eventHandlers = Symbol('_eventHandlers');
exports._eventHandlers = _eventHandlers;
const eventHandlers = Symbol('eventHandlers');
exports.eventHandlers = eventHandlers;
const _eventHandled = Symbol('_eventHandled');
exports._eventHandled = _eventHandled;
const eventHandled = Symbol('eventHandled');
exports.eventHandled = eventHandled;
const _handleEventError = Symbol('_handleEventError');
exports._handleEventError = _handleEventError;
const handleEventError = Symbol('handleEventError');
exports.handleEventError = handleEventError;
function canonicalizeOptions(options) {
    return {
        capture: false,
        once: false,
        passive: false,
        ...(typeof options === 'boolean'
            ? { capture: options }
            : typeof options === 'object'
                ? options
                : undefined),
    };
}
exports.canonicalizeOptions = canonicalizeOptions;
/**
 * EventTarget relies on option comparison.
 * Only the capture property is used.
 */
function isEqualOptions(options1, options2) {
    return options1.capture === options2.capture;
}
exports.isEqualOptions = isEqualOptions;
