// @matrixai/async-locks/dist/Lock.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const resources_1 = require("@matrixai/resources");
const Semaphore_1 = __importDefault(require("./Semaphore"));
class Lock {
    semaphore = new Semaphore_1.default(1);
    get count() {
        return this.semaphore.count;
    }
    isLocked() {
        return this.semaphore.isLocked();
    }
    lock(ctx) {
        const acquire = this.semaphore.lock(1, ctx);
        return () => {
            const acquireP = acquire();
            return acquireP.then(([release]) => [release, this], undefined, (signal) => {
                // Propagate cancellation to `acquireP`
                signal.addEventListener('abort', () => {
                    acquireP.cancel(signal.reason);
                }, { once: true });
            });
        };
    }
    waitForUnlock(ctx) {
        return this.semaphore.waitForUnlock(1, ctx);
    }
    withF(...params) {
        const f = params.pop();
        return (0, resources_1.withF)([this.lock(...params)], ([lock]) => f(lock));
    }
    withG(...params) {
        const g = params.pop();
        return (0, resources_1.withG)([this.lock(...params)], ([lock]) => g(lock));
    }
}
exports.default = Lock;
