// @matrixai/async-locks/dist/RWLockWriter.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const resources_1 = require("@matrixai/resources");
const async_cancellable_1 = require("@matrixai/async-cancellable");
const Lock_1 = __importDefault(require("./Lock"));
const utils = __importStar(require("./utils"));
const errors = __importStar(require("./errors"));
/**
 * Write-preferring read write lock
 */
class RWLockWriter {
    readersLock = new Lock_1.default();
    writersLock = new Lock_1.default();
    readersRelease;
    readerCountBlocked = 0;
    acquireReadersLockP;
    _readerCount = 0;
    _writerCount = 0;
    get count() {
        return this.readerCount + this.writerCount;
    }
    get readerCount() {
        return this._readerCount + this.readerCountBlocked;
    }
    get writerCount() {
        return this._writerCount;
    }
    /**
     * Check if locked
     * If passed `type`, it will also check that the active lock is of that type
     */
    isLocked(type) {
        if (type === 'read') {
            return this._writerCount === 0 && this.readersLock.isLocked();
        }
        else if (type === 'write') {
            return this.writersLock.isLocked();
        }
        else {
            return this.readersLock.isLocked() || this.writersLock.isLocked();
        }
    }
    lock(...params) {
        const type = (params.length === 2
            ? params[0]
            : typeof params[0] === 'string'
                ? params[0]
                : undefined) ?? 'write';
        const ctx = params.length === 2
            ? params[1]
            : typeof params[0] !== 'string'
                ? params[0]
                : undefined;
        switch (type) {
            case 'read':
                return this.read(ctx);
            case 'write':
                return this.write(ctx);
        }
    }
    read(ctx) {
        ctx = ctx != null ? { ...ctx } : {};
        return () => {
            return utils.setupTimedCancellable(async (ctx) => {
                if (this._writerCount > 0) {
                    ++this.readerCountBlocked;
                    const waitWritersLockP = this.writersLock.waitForUnlock(ctx);
                    try {
                        await waitWritersLockP;
                    }
                    finally {
                        --this.readerCountBlocked;
                    }
                }
                const readerCount = ++this._readerCount;
                // The first reader locks
                if (readerCount === 1) {
                    const acquireReadersLock = this.readersLock.lock(ctx);
                    this.acquireReadersLockP = acquireReadersLock();
                    try {
                        [this.readersRelease] = await this.acquireReadersLockP;
                    }
                    catch (e) {
                        --this._readerCount;
                        throw e;
                    }
                }
                else {
                    // Without this, the second or later reader will always lock faster
                    // than the first reader. This forces the subsequent readers to always
                    // wait for the first reader to settle, while discarding any errors.
                    await this.acquireReadersLockP.catch(() => { });
                }
                let released = false;
                return [
                    async () => {
                        if (released)
                            return;
                        released = true;
                        const readerCount = --this._readerCount;
                        // The last reader unlocks
                        if (readerCount === 0) {
                            await this.readersRelease();
                        }
                    },
                    this,
                ];
            }, true, Infinity, errors.ErrorAsyncLocksTimeout, ctx, []);
        };
    }
    write(ctx) {
        ctx = ctx != null ? { ...ctx } : {};
        return () => {
            return utils.setupTimedCancellable(async (ctx) => {
                ++this._writerCount;
                const acquireWritersLock = this.writersLock.lock(ctx);
                const acquireWritersLockP = acquireWritersLock();
                let writersRelease;
                try {
                    [writersRelease] = await acquireWritersLockP;
                }
                catch (e) {
                    --this._writerCount;
                    throw e;
                }
                const acquireReadersLock = this.readersLock.lock(ctx);
                const acquireReadersLockP = acquireReadersLock();
                try {
                    [this.readersRelease] = await acquireReadersLockP;
                }
                catch (e) {
                    await writersRelease();
                    --this._writerCount;
                    throw e;
                }
                let released = false;
                return [
                    async () => {
                        if (released)
                            return;
                        released = true;
                        await this.readersRelease();
                        await writersRelease();
                        --this._writerCount;
                    },
                    this,
                ];
            }, true, Infinity, errors.ErrorAsyncLocksTimeout, ctx, []);
        };
    }
    waitForUnlock(ctx) {
        const waitReadersLockP = this.readersLock.waitForUnlock(ctx);
        const waitWritersLockP = this.writersLock.waitForUnlock(ctx);
        return async_cancellable_1.PromiseCancellable.all([waitReadersLockP, waitWritersLockP]).then(() => { }, undefined, (signal) => {
            signal.addEventListener('abort', () => {
                waitReadersLockP.cancel(signal.reason);
                waitWritersLockP.cancel(signal.reason);
            }, { once: true });
        });
    }
    withF(...params) {
        let type;
        if (params.length === 2) {
            type = params.shift();
        }
        else {
            if (typeof params[0] === 'string') {
                type = params.shift();
            }
            else if (typeof params[0] == null) {
                params.shift();
            }
        }
        type = type ?? 'write';
        switch (type) {
            case 'read':
                return this.withReadF(...params);
            case 'write':
                return this.withWriteF(...params);
        }
    }
    withReadF(...params) {
        const f = params.pop();
        return (0, resources_1.withF)([this.read(...params)], ([lock]) => f(lock));
    }
    withWriteF(...params) {
        const f = params.pop();
        return (0, resources_1.withF)([this.write(...params)], ([lock]) => f(lock));
    }
    withG(...params) {
        let type;
        if (params.length === 2) {
            type = params.shift();
        }
        else {
            if (typeof params[0] === 'string') {
                type = params.shift();
            }
            else if (typeof params[0] == null) {
                params.shift();
            }
        }
        type = type ?? 'write';
        switch (type) {
            case 'read':
                return this.withReadG(...params);
            case 'write':
                return this.withWriteG(...params);
        }
    }
    withReadG(...params) {
        const g = params.pop();
        return (0, resources_1.withG)([this.read(...params)], ([lock]) => g(lock));
    }
    withWriteG(...params) {
        const g = params.pop();
        return (0, resources_1.withG)([this.write(...params)], ([lock]) => g(lock));
    }
}
exports.default = RWLockWriter;
